Route the rest of the WebNavigation events from WebFrameLoaderClient to the extension event listeners.

b-weinstein opened the pull request:
#### d2610dc0dedb035d810463243743ee6c12136d2e
<pre>
Route the rest of the WebNavigation events from WebFrameLoaderClient to the extension event listeners.
<a href="https://bugs.webkit.org/show_bug.cgi?id=249683">https://bugs.webkit.org/show_bug.cgi?id=249683</a>
rdar://102820594

Reviewed by Timothy Hatcher.

This patch allows WebFrameLoaderClient to tell web extensions code about the various stages of frame loading.

The flow is:
- WebFrameLoaderClient sends a message to the WebExtensionController using IPC.
- WebExtensionController iterates over all the WebExtensionContexts and:
        - Sends a message to their WebExtensionContextProxy using IPC.

WebFrameLoaderClient::dispatchDidCommitLoad -&gt; onCommitted and onDOMContentLoaded
WebFrameLoaderClient::dispatchDidFinishLoad -&gt; onCompleted
WebFrameLoaderClient::dispatchDidFailLoad -&gt; onErrorOccurred

And then the WebExtensionContextProxy will fire the event on the extension pages that listen to it.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
(WebKit::WebExtensionController::didCommitLoadForFrame): Call into the WebExtensionContextProxy.
(WebKit::WebExtensionController::didFinishLoadForFrame): Ditto.
(WebKit::WebExtensionController::didFailLoadForFrame): Ditto.
* Source/WebKit/UIProcess/Extensions/WebExtensionController.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionController.messages.in:
* Source/WebKit/WebProcess/Extensions/Cocoa/WebExtensionContextProxyCocoa.mm:
(WebKit::WebExtensionContextProxy::dispatchWebNavigationOnCommittedEvent): Iterate the namespace objects and call the
relevant listeners.
(WebKit::WebExtensionContextProxy::dispatchWebNavigationOnDOMContentLoadedEvent): Ditto.
(WebKit::WebExtensionContextProxy::dispatchWebNavigationOnCompletedEvent): Ditto.
(WebKit::WebExtensionContextProxy::dispatchWebNavigationOnErrorOccurredEvent): Ditto.
* Source/WebKit/WebProcess/Extensions/Cocoa/WebExtensionControllerProxyCocoa.mm:
(WebKit::WebExtensionControllerProxy::didCommitLoadForFrame): Send a message to the WebExtensionController in the UI process.
(WebKit::WebExtensionControllerProxy::didFinishLoadForFrame): Ditto.
(WebKit::WebExtensionControllerProxy::didFailLoadForFrame): Ditto.
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.h:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.messages.in:
* Source/WebKit/WebProcess/Extensions/WebExtensionControllerProxy.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp:
(WebKit::WebFrameLoaderClient::dispatchDidCommitLoad): Call into WebExtensionControllerProxy.
(WebKit::WebFrameLoaderClient::dispatchDidFailLoad): Ditto.
(WebKit::WebFrameLoaderClient::dispatchDidFinishLoad): Ditto.

Canonical link: <a href="https://commits.webkit.org/258594@main">https://commits.webkit.org/258594@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1754e635e337024375ba03af10a568ee57520d3c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/102434 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/11568 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/35478 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/111715 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/12546 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/2473 "Built successfully") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/94720 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/109433 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/108214 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/9601 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/92874 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/94720 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/91476 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/24354 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/94720 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/81/builds/5044 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/25777 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/5196 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/2219 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/11220 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/45270 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5897 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/79/builds/6937 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
```diff
@@ -254,6 +254,57 @@
+{
+    WebExtensionContext::EventListenerTypeSet listenerTypes;
+    listenerTypes.add(WebExtensionEventListenerType::WebNavigationOnCommitted);
+    // FIXME: Is there a more precise time to fire the DOMContentLoaded event? This matches Safari's behavior.
```
Does not appear to be one at the WebKit level.

b-weinstein replied:
I'll just get rid of the FIXME.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
```diff
@@ -254,6 +254,57 @@
+        if (!context->hasPermission(frameURL))
+            continue;
+
+        context->fireEvents(listenerTypes, [context, pageID, frameID, frameURL] {
```
Might be able to just do: `fireEvents({ WebExtensionEventListenerType::WebNavigationOnCompleted }, ...`

b-weinstein replied:
That would be cool! I'll try that.

b-weinstein replied:
It looks like in that case I'd use the event type twice.

I'll do the auto suggestion and the inline definition you suggested before.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
```diff
@@ -254,6 +254,57 @@
+
+void WebExtensionController::didFinishLoadForFrame(WebPageProxyIdentifier pageID, WebCore::FrameIdentifier frameID, URL frameURL)
+{
+    WebExtensionContext::EventListenerTypeSet listenerTypes;
```
If you can't do it inline below, then this should work here:

`auto listenerTypes = WebExtensionContext::EventListenerTypeSet { WebExtensionEventListenerType::WebNavigationOnCompleted }`

b-weinstein committed:
Route the rest of the WebNavigation events from WebFrameLoaderClient to the extension event listeners.
https://bugs.webkit.org/show_bug.cgi?id=249683
rdar://102820594

Reviewed by Timothy Hatcher.

This patch allows WebFrameLoaderClient to tell web extensions code about the various stages of frame loading.

The flow is:
- WebFrameLoaderClient sends a message to the WebExtensionController using IPC.
- WebExtensionController iterates over all the WebExtensionContexts and:
        - Sends a message to their WebExtensionContextProxy using IPC.

WebFrameLoaderClient::dispatchDidCommitLoad -> onCommitted and onDOMContentLoaded
WebFrameLoaderClient::dispatchDidFinishLoad -> onCompleted
WebFrameLoaderClient::dispatchDidFailLoad -> onErrorOccurred

And then the WebExtensionContextProxy will fire the event on the extension pages that listen to it.

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
(WebKit::WebExtensionController::didCommitLoadForFrame): Call into the WebExtensionContextProxy.
(WebKit::WebExtensionController::didFinishLoadForFrame): Ditto.
(WebKit::WebExtensionController::didFailLoadForFrame): Ditto.
* Source/WebKit/UIProcess/Extensions/WebExtensionController.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionController.messages.in:
* Source/WebKit/WebProcess/Extensions/Cocoa/WebExtensionContextProxyCocoa.mm:
(WebKit::WebExtensionContextProxy::dispatchWebNavigationOnCommittedEvent): Iterate the namespace objects and call the
relevant listeners.
(WebKit::WebExtensionContextProxy::dispatchWebNavigationOnDOMContentLoadedEvent): Ditto.
(WebKit::WebExtensionContextProxy::dispatchWebNavigationOnCompletedEvent): Ditto.
(WebKit::WebExtensionContextProxy::dispatchWebNavigationOnErrorOccurredEvent): Ditto.
* Source/WebKit/WebProcess/Extensions/Cocoa/WebExtensionControllerProxyCocoa.mm:
(WebKit::WebExtensionControllerProxy::didCommitLoadForFrame): Send a message to the WebExtensionController in the UI process.
(WebKit::WebExtensionControllerProxy::didFinishLoadForFrame): Ditto.
(WebKit::WebExtensionControllerProxy::didFailLoadForFrame): Ditto.
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.h:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.messages.in:
* Source/WebKit/WebProcess/Extensions/WebExtensionControllerProxy.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp:
(WebKit::WebFrameLoaderClient::dispatchDidCommitLoad): Call into WebExtensionControllerProxy.
(WebKit::WebFrameLoaderClient::dispatchDidFailLoad): Ditto.
(WebKit::WebFrameLoaderClient::dispatchDidFinishLoad): Ditto.

Canonical link: https://commits.webkit.org/258594@main
